src/algorithms/piecewise_linear_stretching.jl

```diff
@@ -0,0 +1,179 @@
+imgo = adjust_histogram(img, f)
+```
+
+# References
+1. W. Burger and M. J. Burge. *Digital Image Processing*. Texts in Computer Science, 2016. [doi:10.1007/978-1-4471-6684-9](https://doi.org/10.1007/978-1-4471-6684-9)
+"""
+@with_kw struct PiecewiseLinearStretching{T} <: AbstractHistogramAdjustmentAlgorithm where T <: NTuple{N, Pair{<: Union{Real, AbstractGray}, <: Union{Real, AbstractGray}}} where {N}
+    src_intervals::T = (0.0 => 0.1, 0.1 => 0.9, 0.9 => 1.0)
+    dst_intervals::T = (0.0 => 0.2, 0.2 => 1.0, 1.0 => 1.0)
+    function PiecewiseLinearStretching(src_intervals::T₁, dst_intervals::T₂) where {T₁ <: NTuple{N₁, Pair{<: Union{Real, AbstractGray}, <: Union{Real, AbstractGray}}} where {N₁}, T₂ <: NTuple{N₂, Pair{<: Union{Real, AbstractGray}, <: Union{Real, AbstractGray}}} where {N₂}}  
+        length(src_intervals) == length(dst_intervals) ||  throw(ArgumentError("The dst_intervals must define an interval pair for each consecutive src_interval pair, i.e. length(src_intervals) == length(dst_intervals)."))
+        for src_pair in src_intervals
+            if first(src_pair) == last(src_pair) 
+                throw(ArgumentError("The start and end of an interval in src_intervals cannot be equal."))
+            end
+        end
+        T = promote_type(T₁, T₂)
+        new{T}(src_intervals, dst_intervals)
+    end    
+end
+
+function (f::PiecewiseLinearStretching)(out::GenericGrayImage, img::GenericGrayImage) 
+    # Verify that the specified dst_intervals fall inside the permissible range
+    # of the output type. 
+    T = eltype(out)
+    for (a,b) in f.dst_intervals
+        p = typemin(T)
+        r = typemax(T)
+        if  !(p <= a <= r) || !(p <= b <= r)
+            throw(DomainError("The specified interval [$a, $b] falls outside the permissible range [$p, $r] associated with $T"))
+        end
+    end
+    # A linear mapping of a value x in the interval [A, B] to the interval [a,b] is given by
+    # the formula f(x) = (x-A) * ((b-a)/(B-A)) + a. The operation r * x - o is equivalent to
+    # (x-A) * ((b-a)/(B-A)) + a. Precalculate the r and o so that later on an inner loop
+    # contains only multiplication and addition to get better performance.
+    N = length(f.src_intervals)
+    rs = zeros(Float64, N)
+    os = zeros(Float64, N)
+    for (src, dst) in zip(pairs(f.src_intervals), pairs(f.dst_intervals)) 
+        n, (A,B) = src
+        _, (a,b) = dst   
+        rs[n] = (b - a) / (B - A)
+        os[n] = (A*b - B*a) / (B - A)
+    end
+
+    # Determine which interval the source pixel fall into, and apply the accompanying linear
+    # transformation taking into account NaN values. 
+    @inbounds for p in eachindex(img)
+        val = img[p]
+        if isnan(val)
+            out[p] = val
+        else
+            is_inside_src_intervals = false
+            for (n, (A,B)) in pairs(f.src_intervals)                                 
+                if (A <= val < B) || (A <= val <= B && n == N)       

```

Comment:
If you were using an `Interval` this would just be `val ∈ iv`. But as stated above I'm not sure that's the best user-facing interface anyway. 

From a performance standpoint, you can order them and then use `searchsortedfirst` which does binary search. Probably not relevant for most cases (you probably don't have very many distinct intervals) but it seems "safe."

Reply:
> From a performance standpoint, you can order them and then use searchsortedfirst which does binary search. Probably not relevant for most cases (you probably don't have very many distinct intervals) but it seems "safe."

I couldn't quite get that to work, since `isless` is not defined for open-closed intervals, i.e. (0.2, 0.5] in IntervalSets.jl

Reply:
Sorry, I meant a vector of edges, not of intervals. Though you can pass a custom comparator through the `lt` keyword to `searchsortedfirst`, e.g., `searchsortedfirst(a, x; lt=(p, q) -> p.left < q.left)` or something.